execution-engine/contracts/hdac-system/pop/src/contract_votes.rs

```diff
@@ -0,0 +1,226 @@
+            let mut split_name = name.split('_');
+            if Some("a") != split_name.next() {
+                continue;
+            }
+
+            let to_publickey = |hex_str: &str| -> Result<PublicKey> {
+                if hex_str.len() != 64 {
+                    return Err(Error::VoteKeyDeserializationFailed);
+                }
+                let mut key_bytes = [0u8; 32];
+                let _bytes_written = base16::decode_slice(hex_str, &mut key_bytes)
+                    .map_err(|_| Error::VoteKeyDeserializationFailed)?;
+                debug_assert!(_bytes_written == key_bytes.len());
+                Ok(PublicKey::from(key_bytes))
+            };
+
+            let to_hash = |hex_str: &str| -> Result<Key> {
+                if hex_str.len() != 64 {
+                    return Err(Error::VoteKeyDeserializationFailed);
+                }
+                let mut hash_bytes = [0u8; 32];
+                let _bytes_written = base16::decode_slice(hex_str, &mut hash_bytes)
+                    .map_err(|_| Error::VoteKeyDeserializationFailed)?;
+                debug_assert!(_bytes_written == hash_bytes.len());
+                // TODO: How to convert hash to key
+                Ok(Key::Hash(hash_bytes))
+            };
+
+            let hex_key = split_name
+                .next()
+                .ok_or(Error::VoteKeyDeserializationFailed)?;
+            let dapp_user = to_publickey(hex_key)?;
+
+            let hex_key = split_name
+                .next()
+                .ok_or(Error::VoteKeyDeserializationFailed)?;
+            let dapp_owner = to_hash(hex_key)?;
+
+            let balance = split_name
+                .next()
+                .and_then(|b| U512::from_dec_str(b).ok())
+                .ok_or(Error::VotesDeserializationFailed)?;
+
+            votes.insert(
+                VoteKey {
+                    user: dapp_user,
+                    dapp: dapp_owner,
+                },
+                balance,
+            );
+        }
+
+        Ok(Votes(votes))
+    }
+
+    /// Writes the current stakes to the contract's known urefs.
+    pub fn write(votes: &Votes) {
+        // Encode the stakes as a set of uref names.
+        let mut new_urefs: BTreeSet<String> = votes
+            .0
+            .iter()
+            .map(|(vote_key, balance)| {
+                let to_hex_string = |address: PublicKey| -> String {
+                    let bytes = address.value();
+                    let mut ret = String::with_capacity(64);
+                    for byte in &bytes[..32] {
+                        write!(ret, "{:02x}", byte).expect("Writing to a string cannot fail");
+                    }
+                    ret
+                };
+
+                let to_hex_string_from_hash = |hash: Key| -> String {
+                    let bytes = hash
+                        .into_hash()
+                        .unwrap_or_revert_with(ApiError::Deserialize);
```

Comment:
Error handling suggestion for this context.

```rust
let bytes = hash
    .into_hash()
    .expect("VoteKey serialization cannot fail");
```

Reply:
https://github.com/hdac-io/CasperLabs/pull/14/commits/9464b5de93c7bca32387ba684abb64aee0814411